### PR TITLE
feat: support create IVF_HNSW_PQ index in Python

### PR DESCRIPTION
**Pull request by BubbleCal**

### CODE REVIEW

File: python/python/lance/dataset.py [hunk 1]
```diff
@@ -1282,8 +1282,13 @@ def create_index(
...
         - **ivf_centroids**: K-mean centroids for IVF clustering.
 
-        If ``index_type`` is "DISKANN", then the following parameters are optional:
+        Optional parameters for "HNSW":
+        - **max_level**: the maximum number of levels in the graph.
+        - **m**: the number of edges per node in the graph.
+        - **m_max**: the maximum number of edges per node in the graph.
+        - **ef_construction**: the number of nodes to examine during the construction.
 
+        Optional parameters for "DISKANN":
```
Comment (eddyxu) [inline python/python/lance/dataset.py:1291]: Lets just remove DISKANN

File: python/python/lance/dataset.py [hunk 2]
```diff
@@ -1282,8 +1282,13 @@ def create_index(
         - **max_opq_iterations**: the maximum number of iterations for training OPQ.
         - **ivf_centroids**: K-mean centroids for IVF clustering.
 
-        If ``index_type`` is "DISKANN", then the following parameters are optional:
+        Optional parameters for "HNSW":
```
Comment (eddyxu) [inline python/python/lance/dataset.py:1285]: This is `Optional parameters for IVF_HNSW_PQ` , right?

`HNSW` could be another implementation

File: python/python/lance/dataset.py [hunk 3]
```diff
@@ -1369,11 +1374,12 @@ def create_index(
         kwargs["metric_type"] = metric
 
         index_type = index_type.upper()
-        if index_type not in ["IVF_PQ", "DISKANN"]:
+        if index_type not in ["IVF_PQ", "DISKANN", "IVF_HNSW_PQ"]:
```
Comment (eddyxu) [inline python/python/lance/dataset.py:1377]: Remove DiskANN

File: python/python/lance/dataset.py [hunk 4]
```diff
@@ -1369,11 +1374,12 @@ def create_index(
...
             raise NotImplementedError(
-                f"Only IVF_PQ or DiskANN index_types supported. Got {index_type}"
+                f"Only [IVF_PQ, IVF_HNSW_PQ, DiskANN] index types supported. "
+                f"Got {index_type}"
             )
-        if index_type == "IVF_PQ":
+        if index_type.startswith("IVF"):
```
Comment (eddyxu) [inline python/python/lance/dataset.py:1382]: Do we check PQ related parameters correctly?

File: python/src/dataset.rs [hunk 5]
```diff
@@ -884,103 +886,135 @@ impl Dataset {
...
-                let mut pq_params = PQBuildParams::default();
-                let mut m_type = MetricType::L2;
-                if let Some(kwargs) = kwargs {
-                    if let Some(mt) = kwargs.get_item("metric_type")? {
-                        m_type = MetricType::try_from(mt.to_string().to_lowercase().as_str())
-                            .map_err(|err| PyValueError::new_err(err.to_string()))?;
+        let params: Box<dyn IndexParams> = if index_type == "BTREE" {
+            Box::<ScalarIndexParams>::default()
+        } else {
+            let mut m_type = MetricType::L2;
```
Comment (eddyxu) [inline python/src/dataset.rs:892]: worth extract to a new function?

**Review by eddyxu — COMMENTED** [review]
Minor issues

File: python/python/tests/test_vector_index.py [hunk 6]
```diff
@@ -343,6 +343,18 @@ def test_create_dot_index(dataset, tmp_path):
...
+def test_create_ivf_hnsw_pq_index(dataset, tmp_path):
+    assert not dataset.has_index
+    ann_ds = lance.write_dataset(dataset.to_table(), tmp_path / "indexed.lance")
+    ann_ds = ann_ds.create_index(
+        "vector",
+        index_type="IVF_HNSW_PQ",
+        num_partitions=4,
+        num_sub_vectors=16,
+    )
+    assert ann_ds.has_index
```
Comment (eddyxu) [inline python/python/tests/test_vector_index.py:355]: Can we use some index metadata to verify correct index type?

File: python/python/lance/dataset.py (hunk 4 above)
Comment (BubbleCal) [reply python/python/lance/dataset.py:1377]: we do, divide the checks into two blocks

**Review by eddyxu — APPROVED** [review]

**Pull request merged**